### PR TITLE
[FIRRTL][LowerToHW] Passthrough any operation not using FIRRTL types

### LINKED ISSUE #1838: [FIRRTL] FIRRTLToHW Ignore any operations which do not belong to the FIRRTL dialect
via https://github.com/llvm/circt/pull/1827#issuecomment-924374528

We are ignoring `sv.interface` and `sv.verbatim` operations when encountered during the lowering.  The suggestion is to  pass through all operations, so long as they have no result/operands/region members that have FIRRTL type.  I was originally going to follow up with it right away, but since I am distracted I'll document this here so I don't forget.

### PR DESCRIPTION
**Pull request by youngar**
We had some hardcoded logic to allow `sv.verbatim` and `sv.interface`
operations to be copied from a `CircuitOp` when lowering to HW.
This change makes this more generic by automatically passing through any
operation which (recursively) does not have any FIRRTL type or
operation.

This change only affects operations which appear in the CircuitOp's
body.  Operations appearing in a ModuleBody are handled by
`handleUnloweredOp`.

Fixes #1838

### CODE REVIEW

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 1]
```diff
@@ -80,6 +80,49 @@ static Type lowerType(Type type) {
...
+// operation.  This checks that the operation recursively has no FIRRTL
+// operations or types.
+static LogicalResult verifyOpLegality(Operation *op) {
+  SmallVector<Operation *> worklist = {op};
+  while (!worklist.empty()) {
+    op = worklist.pop_back_val();
+
+    // Check that this operation is not a FIRRTL op.
+    if (op->getDialect() && (FIRRTLDialect::getDialectNamespace() ==
+                             op->getDialect()->getNamespace()))
```
Comment (lattner) [inline lib/Conversion/FIRRTLToHW/LowerToHW.cpp:93]: I'd recommend splitting this to the top of the function for readability:

`StringRef firrtlNamespace = FIRRTLDialect::getDialectNamespace();`

Or just use `op->getDialect()->getNamespace() == "firrtl"`

Comment (lattner) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:93]: Alternatively can you pointer compare `op->getDialect() == &FIRRTLDialect`?

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 2]
```diff
@@ -80,6 +80,49 @@ static Type lowerType(Type type) {
...
+      return op->emitError("Found unhandled FIRRTL operation '")
+             << op->getName() << "'";
+
+    // Helper to check a TypeRange for any FIRRTL types.
+    auto checkTypes = [&](TypeRange types) -> LogicalResult {
+      if (llvm::any_of(types, [](Type type) {
+            return type.getDialect().getNamespace() ==
+                   FIRRTLDialect::getDialectNamespace();
```
Comment (lattner) [inline lib/Conversion/FIRRTLToHW/LowerToHW.cpp:101]: I'd recommend just using `isa<FIRRTLType>(`

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 3]
```diff
@@ -80,6 +80,49 @@ static Type lowerType(Type type) {
   return {};
 }
 
+/// This verifies that the target operation has been lowered to a legal
+// operation.  This checks that the operation recursively has no FIRRTL
+// operations or types.
+static LogicalResult verifyOpLegality(Operation *op) {
+  SmallVector<Operation *> worklist = {op};
+  while (!worklist.empty()) {
```
Comment (lattner) [inline lib/Conversion/FIRRTLToHW/LowerToHW.cpp:88]: Would it be reasonable to just use `walk`?

It is generally ok to write things recursive on the region tree, so this doesn't need to be worklist driven. worklist is also fine tho

**Review by lattner — APPROVED** [review]
cool thanks for generalizing this!

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp (hunk 1 above)
Comment (youngar) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:93]: It looks like I can do `isa<FIRRTLDialect>(op->getDialect())` which compares TypeID instead. We might be able to do the following for a pointer compare, but I think i like the first thing better:
```c++
auto dialect = context->getLoadedDialect<FIRRTLDIalect>();
if (op->getDialect() == firrtlDialect)
```

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp (hunk 2 above)
Comment (youngar) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:101]: So there are some types in the FIRRTL dialect that don't inherit from from `FIRRTLType`, namely the memory port types `cmemory` and `cmemoryport`.  They aren't the sorts of types we're too worried about showing up in other ops, do you think we should protect against this?

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp (hunk 3 above)
Comment (youngar) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:88]: I think walk is way better, I don't want to be adding all ops in some giant function to a worklist..

**Pull request merged**